dracut-module-setup.sh

```diff
@@ -378,6 +378,19 @@ use_ipv4_or_ipv6() {
     fi
 }
 
+# nm-wait-online-initrd.service installed by dracut's 35-networkmanager
+# module call nm-online with "-s" thus it doesn't wait for network connectivity
+# to be established. Remove "-s" to avoid this undesirable effect.
```

Comment:
I am not sure this is right. The `-s` is there intentionally. It means for nm-online (and thus `nm-wait-online-initrd.service` and `NetworkManager-wait-online.service`) to wait until NetworkManager logs a line "startup complete".

That moment can be configured via the profiles. Explained in [the manual](https://networkmanager.dev/docs/api/latest/NetworkManager-wait-online.service.html).

The previous patch even modifies `may-fail` accordingly, which is the right way to ensure that startup-complete is not reached before the respective address family is ready.

I don't think that `nm-online` is useful without `-s` nor useful for anything aside the wait-online services. If you don't pass `-s`, it will wait until the state is `>= LOCAL` ([here](https://gitlab.freedesktop.org/NetworkManager/NetworkManager/-/blob/8a0aa2c8acfd540d5455dd16bbf2bf9b10756418/src/nm-online/nm-online.c#L119)). If you set `ipv6.may-fail=no` (thus, startup-complete is not yet reached), then the state can already be `>= LOCAL` due to IPv4.

Reply:
maybe we could also set `connection.wait-device-timeout=30000` for all the copied profiles, just to be sure.
And, drop a file `/etc/NetworkManager/conf.d/95-kexec.conf` with `[device-95-kexec] carrier-wait-timeout=30000`.

(see `man nm-settings` and `man NetworkManager.conf`, respectively).

Reply:
> I am not sure this is right. The `-s` is there intentionally. It means for nm-online (and thus `nm-wait-online-initrd.service` and `NetworkManager-wait-online.service`) to wait until NetworkManager logs a line "startup complete".

Currently there are a few issues (e.g. https://bugzilla.redhat.com/show_bug.cgi?id=2035451 ) that sending files to a remote fs failed because network is not ready despite it's done after nm-wait-online-initrd.service has successfully returned. According to [the documentation of nm-online](https://networkmanager.dev/docs/api/latest/nm-online.html), `-s | --wait-for-startup` "waits for NetworkManager startup to complete, rather than waiting for network connectivity specifically". So this is not we wants.

> If you don't pass -s, it will wait until the state is >= LOCAL ([here](https://gitlab.freedesktop.org/NetworkManager/NetworkManager/-/blob/8a0aa2c8acfd540d5455dd16bbf2bf9b10756418/src/nm-online/nm-online.c#L119)). 

The provided link points to the last else (L119) branch but with `-s` specified, it seems the `else if (data->wait_startup)` branch is chosen,
```c
    } else if (data->wait_startup) {
        if (!nm_client_get_startup(data->client)) {
            _return(data, EXIT_SUCCESS);
            return TRUE;
        }
    } else {
        if (state == NM_STATE_CONNECTED_LOCAL || state == NM_STATE_CONNECTED_SITE
            || state == NM_STATE_CONNECTED_GLOBAL) {
            _return(data, EXIT_SUCCESS);
            return TRUE;
        }
    }
```

> If you set ipv6.may-fail=no (thus, startup-complete is not yet reached), then the state can already be >= LOCAL due to IPv4.

Thanks for raising this problem! Another problem I noticed after removing `-s` is when I need multiple network interfaces to be read, nm-online returns immediately once even only one network interface is ready. 

> maybe we could also set `connection.wait-device-timeout=30000` for all the copied profiles, just to be sure. And, drop a file `/etc/NetworkManager/conf.d/95-kexec.conf` with `[device-95-kexec] carrier-wait-timeout=30000`.
> 
> (see `man nm-settings` and `man NetworkManager.conf`, respectively).

But the problem with `-s` is nm-online returns when certain timeouts have been reached regardless the network is ready or not and there are many timeout parameters including this device.timeout and other parameters like dhcp.timeout, carrier.timeout and etc. To figure out how these timeout parameters play its role seems to be a difficult task. Another approach I have been thinking today is to implement the logic of waiting for network to be ready in kexec-tools, for example to use `ip -o route get to SRC_IP` or `ping` to tell if the network is ready. But I'm not sure if they are robust. Compared with the timeout approach, which way would you recommend?  For kexec-tools, this dracut module is executed when the network is ready so I think it's safe to assume given enough time, the same connection profiles can be used to bring up the needed connection successfully. But it's a bit tricky for me to figure out the values of these timeout parameters. Btw, nm-intird-generator uses  `connection.wait-device-timeout=60000`.

Reply:
`NetworkManager-wait-online.service` (and `nm-online -s`, and `nm-wait-online-initrd.service`) wait until the service is fully configured according to the profiles. The user is in control (via the profile settings) how long to wait for a device to appear (`connection.wait-device-timeout`), how long to wait before giving up on DHCP (`ipv4.dhcp-timeout`), whether IPv4 is mandatory (`ipv4.may-fail`), etc.

If it unexpectedly completes too early, you can investigate why, and possibly adjust the profile to fix that (unless there is a bug or a missing feature). Granted, in this case we take the profiles that the user configured for different purposes,so they might not have the right settings suitable for initrd. E.g. `connection.wait-device-timeout` might not matter in real-root, but matter in initrd when the driver only get loaded.

So optimally, the profiles do the right thing and "startup-complete" happens at the desired moment.

In any case, if you have your own requirements, it's perfectly fine to replace `nm-wait-online-initrd.service` with your own service that is `Before=network-online.target, Before=dracut-initqueue.service` and implementes the waiting in any way you wish (including `ip route get` or `ping`). You can then disable `nm-wait-online-initrd.service`.

Reply:
> `NetworkManager-wait-online.service` (and `nm-online -s`, and `nm-wait-online-initrd.service`) wait until the service is fully configured according to the profiles. The user is in control (via the profile settings) how long to wait for a device to appear (`connection.wait-device-timeout`), how long to wait before giving up on DHCP (`ipv4.dhcp-timeout`), whether IPv4 is mandatory (`ipv4.may-fail`), etc.
> 
> If it unexpectedly completes too early, you can investigate why, and possibly adjust the profile to fix that (unless there is a bug or a missing feature). Granted, in this case we take the profiles that the user configured for different purposes,so they might not have the right settings suitable for initrd. E.g. `connection.wait-device-timeout` might not matter in real-root, but matter in initrd when the driver only get loaded.
> 
> So optimally, the profiles do the right thing and "startup-complete" happens at the desired moment.
> 
> In any case, if you have your own requirements, it's perfectly fine to replace `nm-wait-online-initrd.service` with your own service that is `Before=network-online.target, Before=dracut-initqueue.service` and implementes the waiting in any way you wish (including `ip route get` or `ping`). You can then disable `nm-wait-online-initrd.service`.

Thanks for sharing your thoughts! Combining your your suggestions with the reasoning the old approach works for most of the cases,  I apply the timeout configuration of nm-initrd-generator with an additional waiting using `ip route get` to make the sure the network is ready before transferring a file to remote target in new version. Other changes includes all the suggestions and issues found by you and some other minor changes,
 - better naming likes `s/clone_nmconnections/clone_and_modify_nmconnections`, `s&/etc/NetworkManager/conf.d/10-kdump-netif.conf&/etc/NetworkManager/conf.d/10-kdump-netif_allowlist.conf`
 - generate a unique list of the names of needed netifs to avoid to writing duplicate strings like `[device-others]
match-device=except:mac:00:02:55:94:21:38;except:mac:00:02:55:94:21:38;` to  `etc/NetworkManager/conf.d/10-kdump-netif_allowlist.conf` and avoid installing network drivers with duplicate names
 - drop unneeded `find_online_znet_device` which was used to check if znet is used